singletons generates ill-kinded type signature for class method

The following code fails to single with `singletons-2.4.1` and HEAD:

```haskell
{-# LANGUAGE GADTs #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE UndecidableInstances #-}
module Bug where

import Data.Kind
import Data.Singletons.TH

$(singletons [d|
  class C (f :: k -> Type) where
    method :: f a
  |])

$(singletons [d|
  instance C [] where
    method = []
  |])
```
```
$ /opt/ghc/8.4.3/bin/ghc Bug.hs -ddump-splices
[1 of 1] Compiling Bug              ( Bug.hs, Bug.o )
Bug.hs:(13,3)-(16,4): Splicing declarations
    singletons
      [d| class C_a2X8 (f_a3qK :: k_a2Xa -> Type) where
            method_a2X9 :: f_a3qK a_a3qL |]
  ======>
    class C_a6d2 (f_a6d5 :: k_a6d4 -> Type) where
      method_a6d3 :: f_a6d5 a_a6d6
    type MethodSym0 = Method
    class PC (f_a6d5 :: k_a6d4 -> Type) where
      type Method :: f_a6d5 a_a6d6
    class SC (f_a6d5 :: k_a6d4 -> Type) where
      sMethod :: Sing (MethodSym0 :: f_a6d5 a_a6d6)
Bug.hs:(18,3)-(21,4): Splicing declarations
    singletons
      [d| instance C [] where
            method = [] |]
  ======>
    instance C GHC.Types.[] where
      method = []
    type family Method_6989586621679034099_a6k4 :: [a_a6d6] where
      Method_6989586621679034099_a6k4 = '[]
    type Method_6989586621679034099Sym0 =
        Method_6989586621679034099_a6k4
    instance PC [] where
      type Method = Method_6989586621679034099Sym0
    instance SC [] where
      sMethod ::
        forall (a_a6d6 :: []). Sing (MethodSym0 :: f_a6d5 a_a6d6)
      sMethod = singletons-2.4.1:Data.Singletons.Prelude.Instances.SNil

Bug.hs:18:3: error:
    • Expecting one more argument to ‘[]’
      Expected a type, but ‘[]’ has kind ‘* -> *’
    • In the kind ‘[]’
      In the type signature:
        sMethod :: forall (a_a6d6 :: []).
                   Sing (MethodSym0 :: f_a6d5 a_a6d6)
      In the instance declaration for ‘SC []’
   |
18 | $(singletons [d|
   |   ^^^^^^^^^^^^^^...
```

Two important things to note:

* This relies on the declaration for `instance C []` instance being in a separate splice from the declaration for `class C f`. In other words, the bug appears to have something to do with the way that `singletons` is reifying `SC`.
* `f` having kind `k -> Type` is also important, because changing it to `Type -> Type` makes the bug go away.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -111,6 +111,8 @@ Changelog for singletons project
 
 * Permit singling of expression and pattern signatures.
 
+* Permit promotion and singling of `InstanceSigs`.
+
 * `sError` and `sUndefined` now have `HasCallStack` constraints, like their
   counterparts `error` and `undefined`. The promoted and singled counterparts
   to `errorWithoutStackTrace` have also been added in case you do not want

--- a/README.md
+++ b/README.md
@@ -592,6 +592,7 @@ The following constructs are fully supported:
 * class and instance declarations
 * scoped type variables
 * signatures (e.g., `(x :: Maybe a)`) in expressions and patterns
+* `InstanceSigs`
 * higher-kinded type variables (see below)
 * finite arithmetic sequences (see below)
 * functional dependencies (with limitations -- see below)
@@ -740,3 +741,24 @@ Known bugs
 * Singled code that contains uses type families is likely to fail due to GHC
   Trac #12564. Note that singling type family declarations themselves is fine
   (and often desired, since that produces defunctionalization symbols for them).
+* Singling instances of poly-kinded type classes is likely to fail due to
+  [#358](https://github.com/goldfirere/singletons/issues/358).
+  However, one can often work around the issue by using `InstanceSigs`. For
+  instance, the following code will not single:
+
+  ```haskell
+  class C (f :: k -> Type) where
+    method :: f a
+
+  instance C [] where
+    method = []
+  ```
+
+  Adding a type signature for `method` in the `C []` is sufficient
+  to work around the issue, though:
+
+  ```haskell
+  instance C [] where
+    method :: [a]
+    method = []
+  ```

--- a/src/Data/Singletons/Deriving/Bounded.hs
+++ b/src/Data/Singletons/Deriving/Bounded.hs
@@ -54,5 +54,6 @@ mkBoundedInstance mb_ctxt ty (DataDecl _ _ cons) = do
   return $ InstDecl { id_cxt = constraints
                     , id_name = boundedName
                     , id_arg_tys = [ty]
+                    , id_sigs  = mempty
                     , id_meths = [ (minBoundName, mk_rhs minRHS)
                                  , (maxBoundName, mk_rhs maxRHS) ] }

--- a/src/Data/Singletons/Deriving/Enum.hs
+++ b/src/Data/Singletons/Deriving/Enum.hs
@@ -50,5 +50,6 @@ mkEnumInstance mb_ctxt ty (DataDecl data_name tvbs cons) = do
                       -- to use Nat instead of Int
 
                    , id_arg_tys = [ty]
+                   , id_sigs    = mempty
                    , id_meths   = [ (singletonsToEnumName, to_enum)
                                   , (singletonsFromEnumName, from_enum) ] })

--- a/src/Data/Singletons/Deriving/Foldable.hs
+++ b/src/Data/Singletons/Deriving/Foldable.hs
@@ -95,4 +95,5 @@ mkFoldableInstance mb_ctxt ty dd@(DataDecl _ _ cons) = do
   return $ InstDecl { id_cxt = constraints
                     , id_name = foldableName
                     , id_arg_tys = [ty]
+                    , id_sigs  = mempty
                     , id_meths = meths }

--- a/src/Data/Singletons/Deriving/Functor.hs
+++ b/src/Data/Singletons/Deriving/Functor.hs
@@ -86,6 +86,7 @@ mkFunctorInstance mb_ctxt ty dd@(DataDecl _ _ cons) = do
   return $ InstDecl { id_cxt = constraints
                     , id_name = functorName
                     , id_arg_tys = [ty]
+                    , id_sigs  = mempty
                     , id_meths = [ (fmapName,    UFunction fmap_clauses)
                                  , (replaceName, UFunction replace_clauses)
                                  ] }

--- a/src/Data/Singletons/Deriving/Ord.hs
+++ b/src/Data/Singletons/Deriving/Ord.hs
@@ -37,6 +37,7 @@ mkOrdInstance mb_ctxt ty (DataDecl _ _ cons) = do
   return (InstDecl { id_cxt = constraints
                    , id_name = ordName
                    , id_arg_tys = [ty]
+                   , id_sigs  = mempty
                    , id_meths = [(compareName, UFunction clauses)] })
 
 mk_equal_clause :: Quasi q => DCon -> q DClause

--- a/src/Data/Singletons/Deriving/Show.hs
+++ b/src/Data/Singletons/Deriving/Show.hs
@@ -34,6 +34,7 @@ mkShowInstance mb_ctxt ty (DataDecl _ _ cons) = do
   return $ InstDecl { id_cxt = constraints
                     , id_name = showName
                     , id_arg_tys = [ty]
+                    , id_sigs  = mempty
                     , id_meths = [ (showsPrecName, UFunction clauses) ] }
 
 mk_showsPrec :: DsMonad q => [DCon] -> q [DClause]

--- a/src/Data/Singletons/Deriving/Traversable.hs
+++ b/src/Data/Singletons/Deriving/Traversable.hs
@@ -65,4 +65,5 @@ mkTraversableInstance mb_ctxt ty dd@(DataDecl _ _ cons) = do
   return $ InstDecl { id_cxt = constraints
                     , id_name = traversableName
                     , id_arg_tys = [ty]
+                    , id_sigs  = mempty
                     , id_meths = [ (traverseName, UFunction trav_clauses) ] }

--- a/src/Data/Singletons/Partition.hs
+++ b/src/Data/Singletons/Partition.hs
@@ -33,6 +33,7 @@ import Language.Haskell.TH.Desugar
 import Data.Singletons.Util
 
 import Control.Monad
+import Data.Bifunctor (bimap)
 import Data.List.NonEmpty (NonEmpty(..))
 import qualified Data.Map as Map
 import Data.Map (Map)
@@ -101,11 +102,13 @@ partitionDec (DClassD cxt name tvbs fds decs) = do
                                                , cd_lde       = lde }]
                   , pd_open_type_family_decs = otfs }
 partitionDec (DInstanceD _ cxt ty decs) = do
-  defns <- liftM catMaybes $ mapM partitionInstanceDec decs
+  (defns, sigs) <- liftM (bimap catMaybes mconcat) $
+                   mapAndUnzipM partitionInstanceDec decs
   (name, tys) <- split_app_tys [] ty
   return $ mempty { pd_instance_decs = [InstDecl { id_cxt       = cxt
                                                  , id_name      = name
                                                  , id_arg_tys   = tys
+                                                 , id_sigs      = sigs
                                                  , id_meths     = defns }] }
   where
     split_app_tys acc (DAppT t1 t2) = split_app_tys (t2:acc) t1
@@ -170,13 +173,20 @@ partitionClassDec (DTySynInstD {}) =
 partitionClassDec _ =
   fail "Only method declarations can be promoted within a class."
 
-partitionInstanceDec :: Monad m => DDec -> m (Maybe (Name, ULetDecRHS))
+partitionInstanceDec :: Monad m => DDec
+                     -> m ( Maybe (Name, ULetDecRHS) -- right-hand sides of methods
+                          , Map Name DType           -- method type signatures
+                          )
 partitionInstanceDec (DLetDec (DValD (DVarPa name) exp)) =
-  return $ Just (name, UValue exp)
+  pure (Just (name, UValue exp), mempty)
 partitionInstanceDec (DLetDec (DFunD name clauses)) =
-  return $ Just (name, UFunction clauses)
-partitionInstanceDec (DLetDec (DPragmaD {})) = return Nothing
-partitionInstanceDec (DTySynInstD {}) = pure Nothing
+  pure (Just (name, UFunction clauses), mempty)
+partitionInstanceDec (DLetDec (DSigD name ty)) =
+  pure (Nothing, Map.singleton name ty)
+partitionInstanceDec (DLetDec (DPragmaD {})) =
+  pure (Nothing, mempty)
+partitionInstanceDec (DTySynInstD {}) =
+  pure (Nothing, mempty)
   -- There's no need to track associated type family instances, since
   -- we already record the type family itself separately.
 partitionInstanceDec _ =
@@ -214,6 +224,7 @@ partitionDeriving mb_strat deriv_pred mb_ctxt ty data_decl =
 
                     , id_name      = deriv_name
                     , id_arg_tys   = arg_tys ++ [ty]
+                    , id_sigs      = mempty
                     , id_meths     = [] }
 
        | Just DNewtypeStrategy <- mb_strat

--- a/src/Data/Singletons/Promote.hs
+++ b/src/Data/Singletons/Promote.hs
@@ -231,8 +231,8 @@ promoteDecs raw_decls = do
     -- promoteLetDecs returns LetBinds, which we don't need at top level
   _ <- promoteLetDecs noPrefix let_decs
   mapM_ promoteClassDec classes
-  let all_meth_sigs = foldMap (lde_types . cd_lde) classes
-  mapM_ (promoteInstanceDec all_meth_sigs) insts
+  let orig_meth_sigs = foldMap (lde_types . cd_lde) classes
+  mapM_ (promoteInstanceDec orig_meth_sigs) insts
   mapM_ promoteDerivedEqDec   derived_eq_decs
   promoteDataDecs datas
 
@@ -315,7 +315,7 @@ promoteClassDec decl@(ClassDecl { cd_cxt  = cxt
     let defaults_list  = Map.toList defaults
         defaults_names = map fst defaults_list
     (default_decs, ann_rhss, prom_rhss)
-      <- mapAndUnzip3M (promoteMethod Nothing meth_sigs) defaults_list
+      <- mapAndUnzip3M (promoteMethod Map.empty Nothing meth_sigs) defaults_list
 
     let infix_decls' = catMaybes $ map (uncurry promoteInfixDecl)
                                  $ Map.toList infix_decls
@@ -361,16 +361,17 @@ promoteClassDec decl@(ClassDecl { cd_cxt  = cxt
 
 -- returns (unpromoted method name, ALetDecRHS) pairs
 promoteInstanceDec :: Map Name DType -> UInstDecl -> PrM AInstDecl
-promoteInstanceDec meth_sigs
+promoteInstanceDec orig_meth_sigs
                    decl@(InstDecl { id_name     = cls_name
                                   , id_arg_tys  = inst_tys
+                                  , id_sigs     = inst_sigs
                                   , id_meths    = meths }) = do
   cls_tvb_names <- lookup_cls_tvb_names
   inst_kis <- mapM promoteType inst_tys
   let kvs_to_bind = foldMap fvDType inst_kis
   forallBind kvs_to_bind $ do
     let subst = Map.fromList $ zip cls_tvb_names inst_kis
-    (meths', ann_rhss, _) <- mapAndUnzip3M (promoteMethod (Just subst) meth_sigs) meths
+    (meths', ann_rhss, _) <- mapAndUnzip3M (promoteMethod inst_sigs (Just subst) orig_meth_sigs) meths
     emitDecs [DInstanceD Nothing [] (foldType (DConT pClsName)
                                       inst_kis) meths']
     return (decl { id_meths = zip (map fst meths) ann_rhss })
@@ -429,28 +430,21 @@ dsReifyTypeNameInfo, which first calls lookupTypeName (to ensure we can find a N
 that's in the type namespace) and _then_ reifies it.
 -}
 
-promoteMethod :: Maybe (Map Name DKind)
+promoteMethod :: Map Name DType -- InstanceSigs for methods
+              -> Maybe (Map Name DKind)
                     -- ^ instantiations for class tyvars (Nothing for default decls)
                     --   See Note [Promoted class method kinds]
               -> Map Name DType     -- method types
               -> (Name, ULetDecRHS)
               -> PrM (DDec, ALetDecRHS, DType)
                  -- returns (type instance, ALetDecRHS, promoted RHS)
-promoteMethod m_subst sigs_map (meth_name, meth_rhs) = do
-  (arg_kis, res_ki) <- lookup_meth_ty
+promoteMethod inst_sigs_map m_subst orig_sigs_map (meth_name, meth_rhs) = do
+  (meth_arg_kis, meth_res_ki) <- lookup_meth_ty
   ((_, _, _, eqns), _defuns, ann_rhs)
-    <- promoteLetDecRHS (Just (arg_kis, res_ki)) sigs_map Map.empty
+    <- promoteLetDecRHS (Just (meth_arg_kis, meth_res_ki)) Map.empty Map.empty
                         noPrefix meth_name meth_rhs
-  meth_arg_tvs <- mapM (const $ qNewName "a") arg_kis
-  let -- If we're dealing with an associated type family instance, substitute
-      -- in the kind of the instance for better kind information in the RHS
-      -- helper function. If we're dealing with a default family implementation
-      -- (m_subst = Nothing), there's no need for a substitution.
-      -- See Note [Promoted class method kinds]
-      do_subst      = maybe id substKind m_subst
-      meth_arg_kis' = map do_subst arg_kis
-      meth_res_ki'  = do_subst res_ki
-      helperNameBase = case nameBase proName of
+  meth_arg_tvs <- mapM (const $ qNewName "a") meth_arg_kis
+  let helperNameBase = case nameBase proName of
                          first:_ | not (isHsLetter first) -> "TFHelper"
                          alpha                            -> alpha
 
@@ -468,14 +462,14 @@ promoteMethod m_subst sigs_map (meth_name, meth_rhs) = do
       -- strictly necessary, as kind inference can figure them out just as well.
       family_args = map DVarT meth_arg_tvs
   helperName <- newUniqueName helperNameBase
-  let tvbs = zipWith DKindedTV meth_arg_tvs meth_arg_kis'
+  let tvbs = zipWith DKindedTV meth_arg_tvs meth_arg_kis
   emitDecs [DClosedTypeFamilyD (DTypeFamilyHead
                                   helperName
                                   tvbs
-                                  (DKindSig meth_res_ki')
+                                  (DKindSig meth_res_ki)
                                   Nothing)
                                eqns]
-  emitDecsM (defunctionalize helperName Nothing tvbs (Just meth_res_ki'))
+  emitDecsM (defunctionalize helperName Nothing tvbs (Just meth_res_ki))
   return ( DTySynInstD
              proName
              (DTySynEqn family_args
@@ -486,17 +480,36 @@ promoteMethod m_subst sigs_map (meth_name, meth_rhs) = do
     proName = promoteValNameLhs meth_name
 
     lookup_meth_ty :: PrM ([DKind], DKind)
-    lookup_meth_ty = case Map.lookup meth_name sigs_map of
-      Nothing -> do
-        mb_info <- dsReifyTypeNameInfo proName
-                   -- See Note [Using dsReifyTypeNameInfo when promoting instances]
-        case mb_info of
-          Just (DTyConI (DOpenTypeFamilyD (DTypeFamilyHead _ tvbs mb_res_ki _)) _)
-            -> let arg_kis = map (default_to_star . extractTvbKind) tvbs
-                   res_ki  = default_to_star (resultSigToMaybeKind mb_res_ki)
-                in return (arg_kis, res_ki)
-          _ -> fail $ "Cannot find type annotation for " ++ show proName
-      Just ty -> promoteUnraveled ty
+    lookup_meth_ty =
+      case Map.lookup meth_name inst_sigs_map of
+        Just ty ->
+          -- We have an InstanceSig. These are easy: no substitution for clas
+          -- variables is required at all!
+          promoteUnraveled ty
+        Nothing -> do
+          -- We don't have an InstanceSig, so we must compute the kind to use
+          -- ourselves (possibly substituting for class variables below).
+          (arg_kis, res_ki) <-
+            case Map.lookup meth_name orig_sigs_map of
+              Nothing -> do
+                mb_info <- dsReifyTypeNameInfo proName
+                           -- See Note [Using dsReifyTypeNameInfo when promoting instances]
+                case mb_info of
+                  Just (DTyConI (DOpenTypeFamilyD (DTypeFamilyHead _ tvbs mb_res_ki _)) _)
+                    -> let arg_kis = map (default_to_star . extractTvbKind) tvbs
+                           res_ki  = default_to_star (resultSigToMaybeKind mb_res_ki)
+                        in return (arg_kis, res_ki)
+                  _ -> fail $ "Cannot find type annotation for " ++ show proName
+              Just ty -> promoteUnraveled ty
+          let -- If we're dealing with an associated type family instance, substitute
+              -- in the kind of the instance for better kind information in the RHS
+              -- helper function. If we're dealing with a default family implementation
+              -- (m_subst = Nothing), there's no need for a substitution.
+              -- See Note [Promoted class method kinds]
+              do_subst      = maybe id substKind m_subst
+              meth_arg_kis' = map do_subst arg_kis
+              meth_res_ki'  = do_subst res_ki
+          pure (meth_arg_kis', meth_res_ki')
 
     default_to_star Nothing  = DConT typeKindName
     default_to_star (Just k) = k

--- a/src/Data/Singletons/Single.hs
+++ b/src/Data/Singletons/Single.hs
@@ -375,8 +375,8 @@ singClassD (ClassDecl { cd_cxt  = cls_cxt
 
 
 singInstD :: AInstDecl -> SgM DDec
-singInstD (InstDecl { id_cxt = cxt, id_name = inst_name
-                    , id_arg_tys = inst_tys, id_meths = ann_meths }) = do
+singInstD (InstDecl { id_cxt = cxt, id_name = inst_name, id_arg_tys = inst_tys
+                    , id_sigs = inst_sigs, id_meths = ann_meths }) = do
   bindContext cxt $ do
     cxt' <- mapM singPred cxt
     inst_kis <- mapM promoteType inst_tys
@@ -393,37 +393,64 @@ singInstD (InstDecl { id_cxt = cxt, id_name = inst_name
     sing_meth name rhs = do
       mb_s_info <- dsReify (singValName name)
       inst_kis <- mapM promoteType inst_tys
-      let mk_subst cls_tvbs = Map.fromList (zip (map extractTvbName cls_tvbs) inst_kis)
-      (s_ty, tyvar_names, ctxt, m_res_ki) <- case mb_s_info of
-        Just (DVarI _ (DForallT cls_tvbs _cls_pred s_ty) _) -> do
-          let subst = mk_subst cls_tvbs
-              (sing_tvbs, ctxt, _args, res_ty) = unravel s_ty
-              m_res_ki = case res_ty of
-                _sing `DAppT` (_prom_func `DSigT` res_ki) -> Just (substKind subst res_ki)
-                _                                         -> Nothing
+      let mk_subst cls_tvbs = Map.fromList $ zip (map extractTvbName vis_cls_tvbs) inst_kis
+            where
+              -- This is a half-hearted attempt to address the underlying problem
+              -- in #358, where we can sometimes have more class type variables
+              -- (due to implicit kind arguments) than class arguments. This just
+              -- ensures that the explicit type variables are properly mapped
+              -- to the class arguments, leaving the implicit kind variables
+              -- unmapped. That could potentially cause *other* problems, but
+              -- those are perhaps best avoided by using InstanceSigs. At the
+              -- very least, this workaround will make error messages slightly
+              -- less confusing.
+              vis_cls_tvbs = drop (length cls_tvbs - length inst_kis) cls_tvbs
 
-          return ( substType subst s_ty
+          sing_meth_ty :: Set Name -> DType
+                       -> SgM (DType, [Name], DCxt, DKind)
+          sing_meth_ty bound_kvs inner_ty = do
+            -- Make sure to expand through type synonyms here! Not doing so
+            -- resulted in #167.
+            raw_ty <- expand inner_ty
+            (s_ty, _num_args, tyvar_names, ctxt, _arg_kis, res_ki)
+              <- singType bound_kvs (promoteValRhs name) raw_ty
+            pure (s_ty, tyvar_names, ctxt, res_ki)
+
+      (s_ty, tyvar_names, ctxt, m_res_ki) <- case Map.lookup name inst_sigs of
+        Just inst_sig -> do
+          -- We have an InstanceSig, so just single that type. Take care to
+          -- avoid binding the variables bound by the instance head as well.
+          let inst_bound = foldMap (fvDType . predToType) cxt <> foldMap fvDType inst_kis
+          (s_ty, tyvar_names, ctxt, res_ki) <- sing_meth_ty inst_bound inst_sig
+          pure (s_ty, tyvar_names, ctxt, Just res_ki)
+        Nothing -> case mb_s_info of
+          -- We don't have an InstanceSig, so we must compute the type to use
+          -- in the singled instance ourselves through reification.
+          Just (DVarI _ (DForallT cls_tvbs _cls_pred s_ty) _) -> do
+            let subst = mk_subst cls_tvbs
+                (sing_tvbs, ctxt, _args, res_ty) = unravel s_ty
+                m_res_ki = case res_ty of
+                  _sing `DAppT` (_prom_func `DSigT` res_ki) -> Just (substKind subst res_ki)
+                  _                                         -> Nothing
+
+            pure ( substType subst s_ty
                  , map extractTvbName sing_tvbs
                  , map (substPred subst) ctxt
                  , m_res_ki )
-        _ -> do
-          mb_info <- dsReify name
-          case mb_info of
-            Just (DVarI _ (DForallT cls_tvbs _cls_pred inner_ty) _) -> do
-              let subst = mk_subst cls_tvbs
-                  cls_kvb_names = foldMap (foldMap fvDType . extractTvbKind) cls_tvbs
-                  cls_tvb_names = Set.fromList $ map extractTvbName cls_tvbs
-                  cls_bound     = cls_kvb_names `Set.union` cls_tvb_names
-              -- Make sure to expand through type synonyms here! Not doing so
-              -- resulted in #167.
-              raw_ty <- expand inner_ty
-              (s_ty, _num_args, tyvar_names, ctxt, _arg_kis, res_ki)
-                <- singType cls_bound (promoteValRhs name) raw_ty
-              return ( substType subst s_ty
+          _ -> do
+            mb_info <- dsReify name
+            case mb_info of
+              Just (DVarI _ (DForallT cls_tvbs _cls_pred inner_ty) _) -> do
+                let subst = mk_subst cls_tvbs
+                    cls_kvb_names = foldMap (foldMap fvDType . extractTvbKind) cls_tvbs
+                    cls_tvb_names = Set.fromList $ map extractTvbName cls_tvbs
+                    cls_bound     = cls_kvb_names `Set.union` cls_tvb_names
+                (s_ty, tyvar_names, ctxt, res_ki) <- sing_meth_ty cls_bound inner_ty
+                pure ( substType subst s_ty
                      , tyvar_names
                      , ctxt
                      , Just (substKind subst res_ki) )
-            _ -> fail $ "Cannot find type of method " ++ show name
+              _ -> fail $ "Cannot find type of method " ++ show name
 
       let kind_map = maybe Map.empty (Map.singleton name) m_res_ki
       meth' <- singLetDecRHS (Map.singleton name tyvar_names)

--- a/src/Data/Singletons/Syntax.hs
+++ b/src/Data/Singletons/Syntax.hs
@@ -73,6 +73,7 @@ data ClassDecl ann = ClassDecl { cd_cxt  :: DCxt
 data InstDecl  ann = InstDecl { id_cxt     :: DCxt
                               , id_name    :: Name
                               , id_arg_tys :: [DType]
+                              , id_sigs    :: Map Name DType
                               , id_meths   :: [(Name, LetDecRHS ann)] }
 
 type UClassDecl = ClassDecl Unannotated

--- a/tests/SingletonsTestSuite.hs
+++ b/tests/SingletonsTestSuite.hs
@@ -101,6 +101,7 @@ tests =
     , compileAndDumpStdTest "T342"
     , compileAndDumpStdTest "FunctorLikeDeriving"
     , compileAndDumpStdTest "T353"
+    , compileAndDumpStdTest "T358"
     ],
     testCompileAndDumpGroup "Promote"
     [ compileAndDumpStdTest "Constructors"

--- a/tests/compile-and-dump/Singletons/T358.ghc86.template
+++ b/tests/compile-and-dump/Singletons/T358.ghc86.template
@@ -1,0 +1,116 @@
+Singletons/T358.hs:(0,0)-(0,0): Splicing declarations
+    singletons
+      [d| class C1 (f :: k -> Type) where
+            method1 :: f a
+          class C2 a where
+            method2a, method2b :: forall b. b -> a
+          
+          instance C1 [] where
+            method1 :: [a]
+            method1 = []
+          instance C2 [a] where
+            method2a _ = []
+            method2b :: forall b. b -> [a]
+            method2b _ = [] |]
+  ======>
+    class C1 (f :: k -> Type) where
+      method1 :: f a
+    instance C1 [] where
+      method1 :: [a]
+      method1 = []
+    class C2 a where
+      method2a :: forall b. b -> a
+      method2b :: forall b. b -> a
+    instance C2 [a] where
+      method2b :: forall b. b -> [a]
+      method2a _ = []
+      method2b _ = []
+    type Method1Sym0 = Method1
+    class PC1 (f :: k -> Type) where
+      type Method1 :: f a
+    type Method2aSym1 (arg0123456789876543210 :: b0123456789876543210) =
+        Method2a arg0123456789876543210
+    instance SuppressUnusedWarnings Method2aSym0 where
+      suppressUnusedWarnings = snd (((,) Method2aSym0KindInference) ())
+    data Method2aSym0 :: forall a0123456789876543210
+                                b0123456789876543210.
+                         (~>) b0123456789876543210 a0123456789876543210
+      where
+        Method2aSym0KindInference :: forall arg0123456789876543210
+                                            arg. SameKind (Apply Method2aSym0 arg) (Method2aSym1 arg) =>
+                                     Method2aSym0 arg0123456789876543210
+    type instance Apply Method2aSym0 arg0123456789876543210 = Method2a arg0123456789876543210
+    type Method2bSym1 (arg0123456789876543210 :: b0123456789876543210) =
+        Method2b arg0123456789876543210
+    instance SuppressUnusedWarnings Method2bSym0 where
+      suppressUnusedWarnings = snd (((,) Method2bSym0KindInference) ())
+    data Method2bSym0 :: forall a0123456789876543210
+                                b0123456789876543210.
+                         (~>) b0123456789876543210 a0123456789876543210
+      where
+        Method2bSym0KindInference :: forall arg0123456789876543210
+                                            arg. SameKind (Apply Method2bSym0 arg) (Method2bSym1 arg) =>
+                                     Method2bSym0 arg0123456789876543210
+    type instance Apply Method2bSym0 arg0123456789876543210 = Method2b arg0123456789876543210
+    class PC2 (a :: Type) where
+      type Method2a (arg :: b) :: a
+      type Method2b (arg :: b) :: a
+    type family Method1_0123456789876543210 :: [a] where
+      Method1_0123456789876543210 = '[]
+    type Method1_0123456789876543210Sym0 = Method1_0123456789876543210
+    instance PC1 [] where
+      type Method1 = Method1_0123456789876543210Sym0
+    type family Method2a_0123456789876543210 (a :: b) :: [a] where
+      Method2a_0123456789876543210 _ = '[]
+    type Method2a_0123456789876543210Sym1 (a0123456789876543210 :: b0123456789876543210) =
+        Method2a_0123456789876543210 a0123456789876543210
+    instance SuppressUnusedWarnings Method2a_0123456789876543210Sym0 where
+      suppressUnusedWarnings
+        = snd (((,) Method2a_0123456789876543210Sym0KindInference) ())
+    data Method2a_0123456789876543210Sym0 :: forall a0123456789876543210
+                                                    b0123456789876543210.
+                                             (~>) b0123456789876543210 [a0123456789876543210]
+      where
+        Method2a_0123456789876543210Sym0KindInference :: forall a0123456789876543210
+                                                                arg. SameKind (Apply Method2a_0123456789876543210Sym0 arg) (Method2a_0123456789876543210Sym1 arg) =>
+                                                         Method2a_0123456789876543210Sym0 a0123456789876543210
+    type instance Apply Method2a_0123456789876543210Sym0 a0123456789876543210 = Method2a_0123456789876543210 a0123456789876543210
+    type family Method2b_0123456789876543210 (a :: b) :: [a] where
+      Method2b_0123456789876543210 _ = '[]
+    type Method2b_0123456789876543210Sym1 (a0123456789876543210 :: b0123456789876543210) =
+        Method2b_0123456789876543210 a0123456789876543210
+    instance SuppressUnusedWarnings Method2b_0123456789876543210Sym0 where
+      suppressUnusedWarnings
+        = snd (((,) Method2b_0123456789876543210Sym0KindInference) ())
+    data Method2b_0123456789876543210Sym0 :: forall a0123456789876543210
+                                                    b0123456789876543210.
+                                             (~>) b0123456789876543210 [a0123456789876543210]
+      where
+        Method2b_0123456789876543210Sym0KindInference :: forall a0123456789876543210
+                                                                arg. SameKind (Apply Method2b_0123456789876543210Sym0 arg) (Method2b_0123456789876543210Sym1 arg) =>
+                                                         Method2b_0123456789876543210Sym0 a0123456789876543210
+    type instance Apply Method2b_0123456789876543210Sym0 a0123456789876543210 = Method2b_0123456789876543210 a0123456789876543210
+    instance PC2 [a] where
+      type Method2a a = Apply Method2a_0123456789876543210Sym0 a
+      type Method2b a = Apply Method2b_0123456789876543210Sym0 a
+    class SC1 (f :: k -> Type) where
+      sMethod1 :: forall a. Sing (Method1Sym0 :: f a)
+    class SC2 a where
+      sMethod2a ::
+        forall b (t :: b). Sing t -> Sing (Apply Method2aSym0 t :: a)
+      sMethod2b ::
+        forall b (t :: b). Sing t -> Sing (Apply Method2bSym0 t :: a)
+    instance SC1 [] where
+      sMethod1 :: forall a. Sing (Method1Sym0 :: [a])
+      sMethod1 = Data.Singletons.Prelude.Instances.SNil
+    instance SC2 [a] where
+      sMethod2a ::
+        forall b (t :: b). Sing t -> Sing (Apply Method2aSym0 t :: [a])
+      sMethod2b ::
+        forall b (t :: b). Sing t -> Sing (Apply Method2bSym0 t :: [a])
+      sMethod2a _ = Data.Singletons.Prelude.Instances.SNil
+      sMethod2b _ = Data.Singletons.Prelude.Instances.SNil
+    instance SC2 a => SingI (Method2aSym0 :: (~>) b a) where
+      sing = (singFun1 @Method2aSym0) sMethod2a
+    instance SC2 a => SingI (Method2bSym0 :: (~>) b a) where
+      sing = (singFun1 @Method2bSym0) sMethod2b

--- a/tests/compile-and-dump/Singletons/T358.hs
+++ b/tests/compile-and-dump/Singletons/T358.hs
@@ -1,0 +1,24 @@
+module T358 where
+
+import Data.Kind
+import Data.Singletons.TH
+
+$(singletons [d|
+  class C1 (f :: k -> Type) where
+    method1 :: f a
+
+  instance C1 [] where
+    method1 :: [a]
+    method1 = []
+
+  class C2 a where
+    method2a, method2b :: forall b. b -> a
+
+  -- Test that variables bound by instance head aren't quantified by the
+  -- generated InstanceSigs
+  instance C2 [a] where
+    method2a _ = []
+
+    method2b :: forall b. b -> [a]
+    method2b _ = []
+  |])

